### DISCUSSION TITLE
Implementing a Metric and including a nn.Module doesn't work correctly in parallel

### DISCUSSION THREAD

**Discussion by import-antigravity** [Lightning-AI/pytorch-lightning#6693] [DDP / multi-GPU / multi-node]
I implemented the FID metric, which involves using a pre-trained Inception network. I have the following code to move it to CUDA:

```python
def __init__(self, ...):
    ...
    self._model = InceptionV3(...)
    if cuda.is_available():
        self._model.to('cuda')
```

When I train using more than one GPU in DDP, this causes an exception `RuntimeError: CUDA error: all CUDA-capable devices are busy or unavailable`. I'm not sure what's causing this. I know that this stuff is supposed to be taken care of automatically but for some reason it's not working for me.

**Comment by import-antigravity** [comment]
Bump

**Reply by jhschwartz** [reply]
Did you ever resolve this? Having the same problem.

**Comment by import-antigravity** [comment]
Sidenote: using `self.to()` causes the same error

**Comment by SkafteNicki** [comment]
Are you using a shared cluster/machine by any chance? That error can be due to another user using the gpu resources (and the gpus set to exclusive mode)

**Reply by import-antigravity** [reply]
Yes I'm using a SLURM cluster. I believe it's not exclusive unless you use the `--exclusive` flag

**Reply by import-antigravity** [reply]
To clarify, I don't think it's due to another user trying to access the resources. This issue only happens when I try to use this metric in a parallel training situation.

**Reply by SkafteNicki** [reply]
Normally, in lightning you should not have to move modules to devices as we do this automatically for you. What happens if you remove the 
```python
if cuda.is_available():
        self._model.to('cuda')
```
part of your code?

**Reply by import-antigravity** [reply]
If I don’t move the module manually, the module does not get moved by the trainer and I get an error. The trainer really doesn’t have any way of controlling the Metric because it’s part of a callback so I’m not sure how that would work. Also I know that Metric extends nn.Module, so in the current version of my code that line is `self.to('cuda')`.

**Reply by SkafteNicki** [reply]
alright, but since you are running in multi-gpu setting, calling `self.to('cuda')` is not correct as it instead should be `self.to(cuda:0)`, `self.to(cuda:1)` ect depending on what gpu currently are being used

**Reply by import-antigravity** [reply]
Ah interesting. Do you have a documentation link for that so I can learn more?

Alternatively, is there a way I can somehow register the Metric instance with the Trainer so the trainer can actually handle the device management? Thanks!

**Reply by SkafteNicki** [reply]
what I would do in your place was to equip my model with the inception net necessary to calculate `fid` so it gets moved to the correct device during training. Then in my callback I can access it through the `pl_module` argument which most hooks have.
